File: F/825-friends-of-appropriate-ages/main.go
package main

import (
	"fmt"
)

func numFriendRequests(ages []int) int {
	buf := make([]int, 121)
	for i := 0; i < len(ages); i++ {
		buf[ages[i]]++
	}

	sums := make([]int, 121)
	sum := 0
	for i := 0; i < len(sums); i++ {
		sum += buf[i]
		sums[i] = sum
	}

	ret := 0
	for i := 0; i < len(ages); i++ {
		age := ages[i]
		min := (age >> 1) + 7
		max := age
		if min < max {
			ret += sums[max] - sums[min] - 1
		}
	}
	return ret
}

func main() {
	fmt.Println("a")
}

File: F/825-friends-of-appropriate-ages/main.rs
struct Solution {}

impl Solution {
  pub fn num_friend_requests(ages: Vec<i32>) -> i32 {
    let mut dp: Vec<i32> = vec![0; 121];
    let mut sum: Vec<i32> = vec![0; 121];
    ages.iter().for_each(|&age| {
      dp[age as usize] += 1;
    });
    (1..121).for_each(|idx| {
      sum[idx] = sum[idx - 1] + dp[idx];
    });
    let mut ans: i32 = 0;
    ages.iter().for_each(|&age| {
      let min = (age >> 1) + 7;
      if min < age {
        ans += sum[age as usize] - sum[min as usize] - 1;
      }
    });
    ans
  }
}
